File: src/chain-observers/classifiers/decline-voting-rights-classifier.ts
import { CollectorClassifierBase, TRegisterEvaluationContext } from "./collector-classifier-base";

export interface IDeclinedVotingRightsAccount {
  account: string;
  effectiveDate: Date;
}

export interface IDeclineVotingRightsAccountsData {
  declineVotingRightsAccounts: Record<string, IDeclinedVotingRightsAccount>;
}

export interface IDeclineVotingRightsCollectorOptions {
  declineVotingRightsAccount: string;
}

export class DeclineVotingRightsClassifier extends CollectorClassifierBase<IDeclineVotingRightsAccountsData, void, void, IDeclineVotingRightsCollectorOptions> {
  public static forOptions(options: IDeclineVotingRightsCollectorOptions): TRegisterEvaluationContext {
    return {
      class: this, // Intentionally using `this` to refer to the class prototype itself later - even though it is not a class **instance**
      options
    };
  }
}

File: python/workerbee/chain_observers/classifiers/decline_voting_rights_classifier.py
from __future__ import annotations

from typing import Any

from ..classifier_results import DeclineVotingRightsData
from .collector_classifier_base import CollectorClassifierBase, TRegisterEvaluationContext


class DeclineVotingRightsClassifier(CollectorClassifierBase[DeclineVotingRightsData]):
    """IDeclineVotingRightsAccountsData:
      decline_voting_rights_accounts: dict[str, IDeclinedVotingRightsAccount]

    IDeclinedVotingRightsAccount:
      account: str
      effective_date: datetime
    """

    @classmethod
    def for_options(cls, options: dict[str, Any]) -> TRegisterEvaluationContext:
        return {"class": cls, "options": options}
